pkg/controller/pxc/users.go

```diff
@@ -19,127 +18,215 @@ import (
+	err = r.client.Get(context.TODO(),
 		types.NamespacedName{
 			Namespace: cr.Namespace,
 			Name:      cr.Spec.SecretsName,
 		},
 		&sysUsersSecretObj,
 	)
 	if err != nil && k8serrors.IsNotFound(err) {
-		return nil
+		return nil, nil, nil
 	} else if err != nil {
-		return errors.Wrapf(err, "get sys users secret '%s'", cr.Spec.SecretsName)
+		return nil, nil, errors.Wrapf(err, "get sys users secret '%s'", cr.Spec.SecretsName)
 	}
 
-	newSysData, err := json.Marshal(sysUsersSecretObj.Data)
-	if err != nil {
-		return errors.Wrap(err, "marshal sys secret data")
+	secretName := internalPrefix + cr.Name
+
+	internalSysSecretObj := corev1.Secret{}
+
+	err = r.client.Get(context.TODO(),
+		types.NamespacedName{
+			Namespace: cr.Namespace,
+			Name:      secretName,
+		},
+		&internalSysSecretObj,
+	)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return nil, nil, errors.Wrap(err, "get internal sys users secret")
+	}
+
+	if k8serrors.IsNotFound(err) {
+		internalSysUsersSecret := sysUsersSecretObj.DeepCopy()
+		internalSysUsersSecret.ObjectMeta = metav1.ObjectMeta{
+			Name:      secretName,
+			Namespace: cr.Namespace,
+		}
+		err = r.client.Create(context.TODO(), internalSysUsersSecret)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "create internal sys users secret")
+		}
+		return nil, nil, nil
 	}
-	newSecretDataHash := sha256Hash(newSysData)
 
-	internalSysSecretObj, err := r.getInternalSysUsersSecret(cr, &sysUsersSecretObj)
+	if cr.Status.PXC.Ready > 0 {
+		err := r.manageOperatorAdminUser(cr, &sysUsersSecretObj, &internalSysSecretObj)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "manage operator admin user")
+		}
+	}
+
+	if cr.Status.Status != api.AppStateReady {
+		return nil, nil, nil
+	}
+
+	newSysData, err := json.Marshal(sysUsersSecretObj.Data)
 	if err != nil {
-		return errors.Wrap(err, "get internal sys users secret")
+		return nil, nil, errors.Wrap(err, "marshal sys secret data")
 	}
+	newSecretDataHash := sha256Hash(newSysData)
 
 	dataChanged, err := sysUsersSecretDataChanged(newSecretDataHash, &internalSysSecretObj)
 	if err != nil {
-		return errors.Wrap(err, "check sys users data changes")
+		return nil, nil, errors.Wrap(err, "check sys users data changes")
 	}
 
 	if !dataChanged {
-		return nil
+		return nil, nil, nil
 	}
 
 	restartPXC, restartProxy, err := r.manageSysUsers(cr, &sysUsersSecretObj, &internalSysSecretObj)
 	if err != nil {
-		return errors.Wrap(err, "manage sys users")
+		return nil, nil, errors.Wrap(err, "manage sys users")
 	}
 
-	err = r.updateInternalSysUsersSecret(cr, &sysUsersSecretObj)
+	internalSysSecretObj.Data = sysUsersSecretObj.Data
+	err = r.client.Update(context.TODO(), &internalSysSecretObj)
 	if err != nil {
-		return errors.Wrap(err, "update internal sys users secret")
+		return nil, nil, errors.Wrap(err, "update internal sys users secret")
 	}
+	pxcAnnotations = make(map[string]string)
+	proxysqlAnnotations = make(map[string]string)
```

Comment:
you always return `nil`-s if pxcAnnotations or proxysqlAnnotations wasn't set. but after that lines, you are going return not nil but an empty map in such cases. Looks inconsistent.

Reply:
https://github.com/percona/percona-xtradb-cluster-operator/pull/495/commits/e9e6f9cec3e7892d1d7184bde1f4a40d648159ba